src/blobstorage.js

```diff
@@ -0,0 +1,253 @@
+import Container from './container';
+import Ajv from 'ajv';
+
+/**
+ * The name of the blob where the JSON schema is stored
+ * @const
+ */
+const JSON_SCHEMA_BLOB_NAME = '.schema.blob.json';
+
+/**
+ * The base name of the JSON schema used created the schema identifier
+ * @const
+ */
+const JSON_SCHEMA_NAME_BASE_URL = 'http://schemas.taskcluster.net/';
+
+/**
+ * Base class for azure blob storage
+ */
+class BlobStorage {
+
+  /**
+   * In case you have the Azure credentials, the options are:
+   * ```js
+   * {
+   *    // Azure credentials
+   *    credentials: {
+   *      accountName: '...',       // Azure account name
+   *      accessKey: '...',         // Azure account key
+   *    }
+   * }
+   * ```
+   *
+   * In case you use SAS, the options are:
+   * TODO
+   */
+  constructor(options) {
+    assert(options, 'options must be specified.');
+    assert(options.credentials, '`options.credentials` must be specified.');
+    assert(typeof options.credentials.accountId === 'string',
+      'The `options.accountId` must be specified and must be a string.');
+    if (options.credentials.accessKey) {
+      assert(typeof options.credentials.accessKey === 'string',
+        'If specified, the `options.accessKey` must be a string.');
+    }
+    this.accountId = options.credentials.accountId;
+    this.accessKey = options.credentials.accessKey || undefined;
+    this.blobsvc = new azure.Blob({
+      accountId: this.accountId,
+      accessKey: this.accessKey,
+    });
+    this.validator = Ajv({useDefaults: true, format: 'full', verbose: true, allErrors: true});
+  }
+
+  /**
+   * List all the containers under the blob storage account.
+   * Note that this method handles also the pagination.
+   *
+   * @param {object} options - Options on the form
+   * ```js
+   * {
+   *    prefix: '...'     // Prefix of containers to list (optional)
+   * }
+   * ```
+   * @returns {Array} containers - An array of Container instances
+   */
+  async listContainers(options) {
+    options = options || {};
+
+    let containers = [];
+    let marker;
+    try {
+      do {
+        let result = await this.blobsvc.listContainers({
+          prefix: options.prefix || undefined,
+          marker: marker,
+        });
+
+        marker = result.nextMarker || undefined;
+        result.containers.forEach(container => {
+          containers.push(new Container({
+            blobService: this.blobsvc,
+            accountId: this.accountId,
+            name: container.name,
+            metadata: container.metadata,
+            eTag: container.properties.eTag,
+          }));
+        });
+      } while (marker);
+    } catch (error) {
+      rethrowDebug(`Failed to list containers with error: ${error}`, error);
+    }
+
+    return containers;
+  }
+
+  async _validate(name, schema) {
+    let validate;
+    let schemaId = `${JSON_SCHEMA_NAME_BASE_URL}${name}#`;
+    try {
+      let schemaValidation = this.validator.getSchema(schemaId);
```

Comment:
I think for now we should not load based on the id, unless the id is the one that would match `JSON_SCHEMA_BLOB_NAME`.  Either way is OK for me.